### PR TITLE
Feature/charts

### PR DESCRIPTION
**Pull request by abubakarazeem**
Added Timeline and statistics in dashboard

### CODE REVIEW

**Comment by abubakarazeem** [comment]
Review: cycle 0

File: app/helpers/issue_types_helper.rb [hunk 1]
```diff
@@ -7,7 +7,12 @@ def issue_type_table_row_class(issue_type)
     issue_type.project_id.nil? ? 'global' : "project-#{issue_type.project_id}"
   end
 
-  def issue_type_project_number(issue_type)
-    issue_type.project_id.nil? ? 'NA' : issue_type.project_id.to_s
+  def issue_type_issues_count(issues_count, issue_type)
+    if issues_count.nil?
```
Comment (basil7vals) [inline app/helpers/issue_types_helper.rb]: issues_count.blank?

File: app/helpers/issue_types_helper.rb [hunk 2]
```diff
@@ -7,7 +7,12 @@ def issue_type_table_row_class(issue_type)
     issue_type.project_id.nil? ? 'global' : "project-#{issue_type.project_id}"
```
Comment (basil7vals) [inline app/helpers/issue_types_helper.rb]: use blank?

**Comment by basil7vals** [comment]
Cycle - 1 completed

File: Gemfile [hunk 3]
```diff
@@ -25,6 +25,7 @@ gem 'uglifier', '>= 1.3.0'
 # Custom Gems
 gem 'audited'
 gem 'cancancan', '3.0.1'
+gem 'chartkick'
```
Comment (abubakarazeem) [inline Gemfile]: mention version

File: app/assets/javascripts/dashboard.js [hunk 4]
```diff
@@ -1,25 +1,39 @@
...
+    info: false,
+    paging: false,
+    scrollY: '50vh',
+    scrollCollapse: true,
     oLanguage: {
         sEmptyTable: "No issue found"
     }
   });
 
+  $('a[data-toggle="pill"]').on('shown.bs.tab', function (e) {
```
Comment (abubakarazeem) [inline app/assets/javascripts/dashboard.js]: should be specific

File: app/controllers/issue_types_controller.rb [hunk 5]
```diff
@@ -5,21 +5,13 @@ class IssueTypesController < ApplicationController
   load_and_authorize_resource
 
   def index
+    @issues_count = current_tenant.issues.group(:issue_type_id).count
```
Comment (abubakarazeem) [inline app/controllers/issue_types_controller.rb]: what will be the query?

File: app/controllers/issue_types_controller.rb [hunk 6]
```diff
@@ -67,6 +59,13 @@ def destroy
     end
   end
 
+  def issues
```
Comment (abubakarazeem) [inline app/controllers/issue_types_controller.rb]: should be in issues controller
restful comment

File: app/helpers/issues_helper.rb [hunk 7]
```diff
@@ -30,4 +30,13 @@ def display_user_name(id)
   def badge_class_from_priority(priority)
     BADGES[priority]
   end
+
+  def issue_delete_data_confirm(issue)
+    watchers_count = issue.issue_watchers.count
+    if watchers_count > 0
+      "Are you sure you want to delete this issue? It contains #{watchers_count} watcher".pluralize(watchers_count)
```
Comment (abubakarazeem) [inline app/helpers/issues_helper.rb]: messages should be in en.yml file

File: app/models/ability.rb [hunk 8]
```diff
@@ -38,8 +38,8 @@ def initialize(user, options = {})
       can :create, Issue, company_id: user.company_id
 
       # IssueWatcher
-      can :create_watcher, IssueWatcher, watcher_id: user.id, watcher_type: IssueWatcher::WATCHER_TYPE_USER
-      can :destroy_watcher, IssueWatcher, watcher_id: user.id, watcher_type: IssueWatcher::WATCHER_TYPE_USER
+      can :create_watcher, IssueWatcher, company_id: user.company_id, watcher_id: user.id, watcher_type: IssueWatcher::WATCHER_TYPES[:user]
```
Comment (abubakarazeem) [inline app/models/ability.rb]: restful action names

File: app/models/issue.rb [hunk 9]
```diff
@@ -38,12 +39,11 @@ class Issue < ActiveRecord::Base
 
   has_many   :documents
```
Comment (abubakarazeem) [inline app/models/issue.rb]: remove this line

File: app/models/issue.rb [hunk 10]
```diff
@@ -38,12 +39,11 @@ class Issue < ActiveRecord::Base
...
 
-  has_many   :comments, as: :commentable
-  has_many :documents, as: :documentable
+  has_many   :documents, as: :documentable, dependent: :destroy
   has_many   :comments, as: :commentable, dependent: :destroy
 
   # Polymorphic Watchers
-  has_many   :issue_watchers
+  has_many   :issue_watchers, dependent: :destroy
   # has_many   :watchers, through: :issue_watchers
```
Comment (abubakarazeem) [inline app/models/issue.rb]: delete comments

File: app/models/issue.rb [hunk 11]
```diff
@@ -38,12 +39,11 @@ class Issue < ActiveRecord::Base
...
   has_many   :watcher_users, through: :issue_watchers, source: :watcher,
```
Comment (abubakarazeem) [inline app/models/issue.rb]: source type and class both are needed?

File: app/models/issue_watcher.rb [hunk 12]
```diff
@@ -1,24 +1,22 @@
...
-  WATCHER_TYPE_TEAM = 'Team'.freeze
+  WATECHER_TYPES = [user: 'User', team: 'Team'].freeze
 
   # Adds watcher to database
   def self.add_watcher(params)
     issue = Issue.find_by(id: params[:issue_id])
     begin
       if issue.present?
-        if params[:watcher_type] == WATCHER_TYPE_USER
+        if params[:watcher_type] == WATCHER_TYPES[:user]
```
Comment (abubakarazeem) [inline app/models/issue_watcher.rb]: can we use 'user' string here?

File: app/models/issue_watcher.rb [hunk 13]
```diff
@@ -54,10 +52,10 @@ def self.remove_watcher(params)
   def self.get_watchers_to_add(params)
     issue = Issue.find_by(id: params[:issue_id])
     if issue.present?
-      if params[:watcher_type] == WATCHER_TYPE_USER
+      if params[:watcher_type] == WATCHER_TYPES[:user]
         watchers = User.where('first_name LIKE ?', "%#{sanitize_sql_like(params[:watcher_search])}%")
```
Comment (abubakarazeem) [inline app/models/issue_watcher.rb]: refactor

File: app/models/issue_watcher.rb [hunk 14]
```diff
@@ -69,10 +67,10 @@ def self.get_watchers_to_add(params)
   def self.get_watchers_to_remove(params)
```
Comment (abubakarazeem) [inline app/models/issue_watcher.rb]: instance method

File: app/views/dashboard/index.html.erb [hunk 15]
```diff
@@ -1,18 +1,28 @@
...
+    <a class="btn btn-outline-secondary mr-2" id="pills_charts_tab" data-toggle="pill" href="#pills_charts" role="tab" aria-controls="pills_charts" aria-selected="false">Statistics</a>
+  </li>
 </ul>
 
-
 <div class="tab-content">
-  <div class="tab-pane fade show active" id="pills_projects" role="tabpanel" aria-labelledby="pills_projects_tab">
+  <div class="tab-pane fade show active" id="pills_timeline" role="tabpanel" aria-labelledby="pills_timeline_tab">
+    <p class="h3 my-4">Issues Timeline</p>
+    <%= timeline @assigned_issues.order(:start_date).pluck(:title, :start_date, :due_date), id: 'timeline_chart' ,height: "70vh" %>
```
Comment (abubakarazeem) [inline app/views/dashboard/index.html.erb]: should be done in controller

File: app/views/dashboard/index.html.erb [hunk 16]
```diff
@@ -72,9 +82,30 @@
...
     </table>
   </div>
-</div>
\ No newline at end of file
+
+  <div class="tab-pane fade" id="pills_charts" role="tabpanel" aria-labelledby="pills_charts_tab">
+    <div class="row">
+      <div class="col-6">
+        <p class="h3 my-4">Assigned Issues</p>
+        <% if @assigned_issues.count > 0 %>
```
Comment (abubakarazeem) [inline app/views/dashboard/index.html.erb]: any?

Comment (abubakarazeem) [inline app/views/dashboard/index.html.erb]: count vs size

File: app/views/dashboard/index.html.erb [hunk 17]
```diff
@@ -72,9 +82,30 @@
...
+          <%= pie_chart @assigned_issues.group_by_issue_state.count %>
```
Comment (abubakarazeem) [inline app/views/dashboard/index.html.erb]: activerecord vs array

File: app/helpers/issue_types_helper.rb (hunk 2 above)
Comment (abubakarazeem) [reply app/helpers/issue_types_helper.rb]: project_id is an integer here. so `nil?` method will do the job

File: app/models/issue_watcher.rb (hunk 12 above)
Comment (basil7vals) [reply app/models/issue_watcher.rb]: For accessing hash with strings also,
- WATCHER_TYPES = WATCHER_TYPES.with_indifferent_access

File: app/helpers/issue_types_helper.rb (hunk 1 above)
Comment (abubakarazeem) [reply app/helpers/issue_types_helper.rb]: not checking for empty hash here.. only for nil

**Pull request merged**